### PR TITLE
Wildcard stable diffusion pipeline

### PR DESCRIPTION
**Pull request by shyamsn97**
# Wildcard Stable Diffusion pipeline

Part of #872 and bigger story #841

Following the great examples from https://github.com/jtkelm2/stable-diffusion-webui-1/blob/master/scripts/wildcards.py and https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Custom-Scripts#wildcards, here's a minimal implementation that allows for users to add "wildcards", denoted by `__wildcard__` to prompts that are used as placeholders for randomly sampled values given by either a dictionary or a `.txt` file. For example:

Say we have a prompt:

```
prompt = "__animal__ sitting on a __object__ wearing a __clothing__"
```

We can then define possible values to be sampled for `animal`, `object`, and `clothing`. These can either be from a `.txt` with the same name as the category, ex `animal.txt`, with contents like:

```
dog
cat
mouse
```

The possible values can also be defined / combined by using a dictionary like: `{"animal":["dog", "cat", mouse"]}`.

The actual pipeline works just like `StableDiffusionPipeline`, except the `__call__` method takes in:

`wildcard_files`: list of file paths for wild card replacement
`wildcard_option_dict`: dict with key as `wildcard` and values as a list of possible replacements
`num_prompt_samples`: number of prompts to sample, uniformly sampling wildcards

A full example:
```python
pipe = WildcardStableDiffusionPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4",
    revision="fp16",
    torch_dtype=torch.float16,
)
prompt = "__animal__ sitting on a __object__ wearing a __clothing__"
out = pipe(
    prompt,
    wildcard_option_dict={
        "clothing":["hat", "shirt", "scarf", "beret"]
    },
    wildcard_files=["object.txt", "animal.txt"],
    num_prompt_samples=1
)
```
Will need to add some additional documentation, but adding this as a start. Hopefully the implementation is straightforward and easy to understand, and let me know your thoughts on how I can improve this :smile:

### CODE REVIEW

File: examples/community/wildcard_stable_diffusion.py [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+        if output_type == "pil":
+            image = self.numpy_to_pil(image)
+
+        if not return_dict:
+            return (image, has_nsfw_concept)
+
+        return WildcardStableDiffusionOutput(images=image, nsfw_content_detected=has_nsfw_concept, prompts=prompt)
+
+
+# if __name__ == "__main__":
```
Comment (patrickvonplaten) [inline examples/community/wildcard_stable_diffusion.py:415]: maybe delete this code? :-)

**Review by patrickvonplaten — COMMENTED** [review]
Hey @shyamsn97,

The PR looks great :-) 
One final thing then I'd be more than happy to merge. 

To increase the easy-of-use and discover-ability of your pipeline, could you add one line to this table: https://github.com/huggingface/diffusers/tree/main/examples/community#community-examples in your PR 

and also add one code example in a new markdown section here: https://github.com/huggingface/diffusers/tree/main/examples/community#example-usages

Also happy to do it for you if you could quickly post one usage example in this PR :-)

**Comment by patrickvonplaten** [comment]
Hey @GreenLandisaLie,

No, this makes total sense! If you prefer just adding a function yourself before the pipeline that's great :-) For people that would like to have everything in one place such a community pipeline could still be useful. 
In general, the idea of community pipeline is to let the community share **very** freely a pipeline example with everybody else by adding it here, so we don't have hard requirements on how the community pipelines should look like

**Comment by GreenLandisaLie** [comment]
> In general, the idea of community pipeline is to let the community share **very** freely a pipeline example with everybody else by adding it here, so we don't have hard requirements on how the community pipelines should look like

Makes sense actually. I did notice a bunch of pipelines (through your link) that I wasn't even aware of and came to that realization soon after posting that comment. Decided to leave it there because the ability to force include/exclude certain strings might be a useful addition to this.

**Comment by shyamsn97** [comment]
Fixed the README but may have done some weird merge shenanigans accidentally, let me know if this pr still works @patrickvonplaten

File: examples/community/README.md [hunk 2]
```diff
@@ -264,3 +265,59 @@ plt.imshow(output.images[0])
 This example produces the following image:
 
 ![image](https://user-images.githubusercontent.com/45072645/196901736-77d9c6fc-63ee-4072-90b0-dc8b903d63e3.png)
```
Comment (patrickvonplaten) [inline examples/community/README.md:267]:
```suggestion
![image](https://user-images.githubusercontent.com/45072645/196901736-77d9c6fc-63ee-4072-90b0-dc8b903d63e3.png)

```

**Comment by patrickvonplaten** [comment]
Fixed it ;-) 

Pulling the `main` GitHub branch with `git pull upstream main` often helps in such cases.

**Comment by patrickvonplaten** [comment]
Thanks a lot for the contribution - looks very nice!

**Review by patrickvonplaten — APPROVED** [review]

**Pull request merged**